### PR TITLE
[HOLD] Patch for getting connection IDs by key

### PR DESCRIPTION
**Pull request by gedu**
### Details

Implemented some improvements on the onyx side, all in the details in the issue ticket

### Fixed Issues

$ https://github.com/Expensify/App/issues/35809
PROPOSAL: -

### Tests

Testing, will be mainly use the app, to know that my changes doesn't break anything specially collections

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by mollfpr** [comment]
Sorry @akinwale, I'll review this 🙏

File: patches/react-native-onyx+2.0.1.patch [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+@@ -819,6 +859,14 @@ function disconnect(connectionID, keyToRemoveFromEvictionBlocklist) {
+     if (keyToRemoveFromEvictionBlocklist) {
+         removeFromEvictionBlockList(keyToRemoveFromEvictionBlocklist, connectionID);
+     }
++
++    const subscriber = callbackToStateMapping[connectionID];
++    const keyToDelete = subscriber.key;
++
++    if (onyxKeyToConnectionIDs.has(keyToDelete)) {
++        onyxKeyToConnectionIDs.set(keyToDelete, _.without(onyxKeyToConnectionIDs.get(keyToDelete), connectionID));
```
Comment (mollfpr) [inline patches/react-native-onyx+2.0.1.patch:93]: Failing perf test.

Screenshot 2024-02-10 at 01 05 48

**Comment by mollfpr** [comment]
I locally pulled the latest `main` to the branch and fixed the issue with the `_ is not defined` locally. Also, run `npm run clean` and run the app with reset-cache. 


All the tests below are tested on the Web, on the native platform it's very laggy. It might be because the `console.log` that's showing for a hundred thousand. Also the issues I found already check on the `main` branch, and it's not reproducible.

Settings
✅ Profile
✅ Share code
✅ Preference

Report
✅ Start a chat
✅ Send a message
✅ Right-click to open the context menu
❌ React to a message
❌ Reply in the thread
❌ Edit a message
✅ Copy to clipboard
✅ Delete message
✅ Mark as unread
✅ Open the emoji picker from the context menu
✅ Open an attachment

Chat composer
✅ Send a message
✅ Send a text with an emoji
✅ Send a text with a markdown
❌ Edit message with up arrow key
✅ Press the actions button
✅ Open the emoji picker
✅ Mention

✅ Assign task

Request money
✅ Manual request
✅ Scan request
✅ Distance request

Split bill
✅ Manual request
✅ Scan request
✅ Distance request

✅ Send money

✅ Add attachment

Workspace
✅ Workspace switcher
✅ Create a new workspace
✅ Workspace profile
✅ Add/remove members

Issues

1. Sometimes can’t open an IOU report and task report. The report keeps loading and request to API `OpenReport`. 
2. High memory usage on Web Chrome
3. The App does not send the API request after use for a while and works after refreshing the page.
4. The message that I try edit previously, suddenly opens the edit message composer and it won’t close. Clicking on the report main composer move the cursor to the edit message composer.
5. The LHN doesn’t change the latest message text after sending a message, need to refresh to correctly see the latest message on LHN.
6. Send multiple requests of money in a row (manual then scan receipt), the IOU preview doesn’t show the scanned receipt thumbnail.

cc @gedu @mountiny

**Comment by gedu** [comment]
@mollfpr Cool great testing coverage. 
Saw what it is going on, it is sending as a key eg: `reportActionsDraft_123344`, so it doesn't find that key, because is never saved as connect, I need to do the same extraction that it is done in other part of the code, will add a fix and update the patch

**Comment by gedu** [comment]
Will do a round of tests but it is working now

https://github.com/Expensify/App/assets/1676818/c6dad586-fba6-45f3-a004-6c2f6b25841f

**Comment by mountiny** [comment]
@mollfpr really great testing, thank you

@gedu let us know when this is ready for another look

**Comment by gedu** [comment]
@mountiny the TS is failing because the App is using onyx v.2.0.2 instead of v2.0.4, I pushed the missing type into the patch so this works, I can remove it later, should be ready for testing

**Comment by mountiny** [comment]
@mollfpr Will you be able to give this another round of testing please?

**Comment by mollfpr** [comment]
@mountiny Yup, running another test!

**Comment by roryabraham** [comment]
@gedu why are you using a patch file to test Onyx? You can test Onyx locally w/ live-reload using the technique described in the README [here](https://github.com/Expensify/react-native-onyx?tab=readme-ov-file#development). PRs can be opened in Onyx and tested by C+ and reviewers in the same way, without making it difficult to read and review the diff.

I suppose the ability to run AdHoc builds is nice, but this PR _should not be merged_, so imo it should be marked as a draft or POC

**Comment by gedu** [comment]
> @gedu why are you using a patch file to test Onyx? You can test Onyx locally w/ live-reload using the technique described in the README [here](https://github.com/Expensify/react-native-onyx?tab=readme-ov-file#development). PRs can be opened in Onyx and tested by C+ and reviewers in the same way, without making it difficult to read and review the diff.
> 
> I suppose the ability to run AdHoc builds is nice, but this PR _should not be merged_, so imo it should be marked as a draft or POC

I created the patch because we were testing with a release build and using the Android device, so the live-reload wouldn't work. 
I [asked](https://github.com/Expensify/App/issues/35809#issuecomment-1931872879) the best way to handle this case if should I create a new PR in Onyx etc. I can make it a draft PR and if this works create a PR for Onyx

**Comment by mountiny** [comment]
Yeah we know its not supposed to be merged, but this works well for testing for any regressions, added HOLD so nobody accidentally merges it

@mollfpr have you been able to conclude another round of testing please?

**Comment by mollfpr** [comment]
Sorry for the delay, the test I did previously now **PASS** on the web and desktop. For the native platform, I can run it locally.

Simulator Screenshot - iPhone 15 Pro 17 2 - 2024-02-17 at 02 11 13

**Comment by mountiny** [comment]
@gedu do you ahve any tips for the native platforms?

**Comment by mountiny** [comment]
@mollfpr can you test these builds?

**Comment by mountiny** [comment]
The ios build des not work for me so I assume the Android wont work either

**Comment by gedu** [comment]
Hey, will check native, it should work we did all the tests on Android

**Comment by gedu** [comment]
Locally I can run it, and it works. 
Is it possible that we only send as key '_123212' or '_' ?

**Comment by mountiny** [comment]
@mollfpr can you test this locally?

**Comment by gedu** [comment]
Should I do a PR in the Onyx repository so you can test it with the watch?

**Comment by gedu** [comment]
@mountiny  Created a PR at onyx repository: https://github.com/Expensify/react-native-onyx/pull/520

**Comment by mountiny** [comment]
Thanks, that will be on hold until we move ahead with onyx bump

**Comment by gedu** [comment]
@mountiny Not sure how to handle this one, but we can close it in favor of: https://github.com/Expensify/react-native-onyx/pull/565

**Comment by mountiny** [comment]
Yep, lets close this now, thank you

**Pull request closed without merging**